### PR TITLE
[Diff] Add tab context menu entry to diff 2 tabs

### LINKED ISSUE #3313: [Diff] Add support for comparing two buffers
### Expected behavior

Currently it is possible to compare two files which exist on disk by selecting them in the sidebar and selecting "Diff Files..."
But there is no way to compare the contents of two tabs (which may or may not be saved / exist on disk etc.).

Ideally, having two tabs multi-selected and opening the tab context menu should show a "Diff Tabs..." option.
Having one tab selected and right-clicking on another tab should show a "Diff with current tab..." option.

### Actual behavior

Have to save the files before comparing them.

### Steps to reproduce

1. Open a tab, paste some text
2. Open another tab, paste some similar text
3. Right click the first tab and there's no option to "Diff with current tab..."

### PR DESCRIPTION
**Pull request by keith-hall**
Fixes #3313

### CODE REVIEW

**Comment by jrappen** [comment]
Thanks for this.

* The mixed use of different quotes for strings was unexpected. Could you choose one?
* Might make sense to use f-strings in line 58.
* If you got extra time, maybe you could add typing info since we're already at `Py3.8` for this package.

**Comment by deathaxe** [comment]
None of ST's core/default plugins uses typing so far. I don't see any reason for them here.

**Comment by jwortmann** [comment]
```
Traceback (most recent call last):
  File "C:\Program Files\Sublime Text\Lib\python38\sublime_plugin.py", line 1480, in run_
    return self.run(**args)
  File "C:\Users\jwortmann\AppData\Roaming\Sublime Text\Packages\Diff\diff.py", line 58, in run
    show_diff_output(diff, None, self, os.path.basename(files[1]) + " -> " + os.path.basename(files[0]))
  File "C:\Users\jwortmann\AppData\Roaming\Sublime Text\Packages\Diff\diff.py", line 109, in show_diff_output
    v = win.create_output_panel('unsaved_changes')
AttributeError: 'DiffFilesCommand' object has no attribute 'create_output_panel'
```

This happens for me on this branch if you use the regular *Diff files...* entry from the sidebar context menu. The `show_diff_output` function seems to expect a `sublime.Window` as the third argument, but recieves a `sublime_plugin.WindowCommand` in line 58. If you change `self` to `self.window` there, it still doesn't work and fails on the second argument, which is supposed to be a `sublime.View`, but acutally is `None`.

---

If I set *"diff_changes_to_buffer": true* in the settings, the tab title appears odd for diffs of two existing files from the tab context menu: there is an unexpected "Unsaved Changes:" prefix and a part of the filename is removed if both files start with a similar filename. For example:

Unbenannt

---

In general I like the idea from this PR. The functionality already exists, and I always wondered why diffs are only available for files from the side bar, instead of all views. No more need to install 3rd-party packages for basic diffs :)

**Comment by keith-hall** [comment]
Thanks for the feedback all, it seems I didn't do enough testing after making some quick changes :sweat_smile: 

---

@jrappen 
> * The mixed use of different quotes for strings was unexpected. Could you choose one?

I like to think that double quotes are being used for human-readable/localizable strings, and single quotes for internal stuff, so I made some changes in that direction. But if you want a single/uniform quote type, let's decide which we prefer and switch them all.

---
@jwortmann should be fixed now

**Comment by jrappen** [comment]
I'm just thinking loud, Keith. It was unexpected for **me**.

I use single quotes. Your reasoning seems fine. Whatever works and/or people are used to.

**Review by deathaxe — DISMISSED** [review]
This change definitly adds value to the existing plugin. I've used it every here and then without issues. Not sure if a new context menu item is something @sublimehq should decide about, but I am fine with it as is.

**Review by jrappen — DISMISSED** [review]

**Comment by jwortmann** [comment]
I can confirm that it works well and doesn't raise an error anymore, but I still have two suggestions for small possible improvements:

1. With the "diff_changes_to_buffer" setting activated, there is a leading backslash (on Windows) in the filepaths of the tab title, which is especially noticeable if the files are in the same folder, e.g. `\file1.py -> \file2.py`. I'd prefer how the *Diff Files...* feature from the side bar shows it without the backslash like `file1.py -> file2.py`. So I would suggest
    ```diff
    -          common_path_length = len(os.path.commonpath(view_names))
    +          common_path_length = len(os.path.commonpath(view_names)) + 1
    ```
in line 168.

2. Maybe use something like "View (15)" or "Unsaved view (15)" as the fallback names for the tab title and diff header, instead of "from_file" and "to_file"?
```diff
- def get_name_for_view(view, fallback_name):
-     return view.file_name() or view.name() or fallback_name
+ def get_name_for_view(view):
+     return view.file_name() or view.name() or "Unsaved view ({})".format(view.id())
```

**Review by jrappen — APPROVED** [review]

**Comment by keith-hall** [comment]
@jwortmann done :) I tweaked the common path logic slightly to also handle when the only part in common is `/` i.e. between `/home/user/file` and `/etc/file` - to keep it, and to ensure that on Windows `D:\file` and `C:\file` would not cut out the drive letter as it would if I used your suggestion directly.

**Comment by jwortmann** [comment]
Perfect :ok_hand:

**Review by jwortmann — APPROVED** [review]

**Review by deathaxe — DISMISSED** [review]

**Review by Ultra-Instinct-05 — APPROVED** [review]

**Comment by Ultra-Instinct-05** [comment]
Been using it for sometime & seems solid so far.

**Review by michaelblyons — APPROVED** [review]

**Comment by michaelblyons** [comment]
Merge?

**Comment by deathaxe** [comment]
Waiting for some comment from sublimehq about the new plugin functionality. 

I am fine with it, though I'd be happy to also be able to diff views from different groups.

File: Diff/Tab Context.sublime-menu [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[
+	{ "command": "diff_views", "args": { "group": -1, "index": -1 }, "caption": "Diff With Current View..." },
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/Tab Context.sublime-menu]: Good point, changed :)

File: Diff/diff.py [hunk 2]
```diff
@@ -87,37 +77,107 @@ def run(self, edit):
...
+        view = view.window().views_in_group(kwargs['group'])[kwargs['index']]
+    return view
+
+
+def get_views_from_tab_context(active_view, **kwargs):
+    selected_views = list(filter(lambda view: view, map(lambda sheet: sheet.view(), active_view.window().selected_sheets())))
+    if 'group' in kwargs and 'index' in kwargs:
+        tab_context_view = get_view_from_tab_context(active_view, **kwargs)
+        if tab_context_view.id() not in map(lambda view: view.id(), selected_views):
+            return selected_views + [tab_context_view]
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py]: Well spotted. Fixed.

File: Diff/diff.py [hunk 3]
```diff
@@ -87,37 +77,107 @@ def run(self, edit):
...
+        from_lines = get_lines_for_view(views[0])
+        to_lines = get_lines_for_view(views[1])
+        add_no_eol_warning_if_applicable(from_lines)
+        add_no_eol_warning_if_applicable(to_lines)
+
+        diff = difflib.unified_diff(
+            from_lines,
+            to_lines,
+            fromfile=view_names[0],
+            tofile  =view_names[1]
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py]: Thanks, removed the spacing.

File: Diff/diff.py [hunk 4]
```diff
@@ -87,37 +77,107 @@ def run(self, edit):
...
-            win.run_command("show_panel", {"panel": "output.unsaved_changes"})
+    use_buffer = not view or view.settings().get('diff_changes_to_buffer')
 
-    def is_enabled(self):
-        return self.view.is_dirty() and self.view.file_name() is not None
+    if use_buffer:
+        v = win.new_file()
+        v.set_name(name)
+        v.set_scratch(True)
+        v.assign_syntax('Packages/Diff/Diff.sublime-syntax')
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py]: done

File: Diff/diff.py [hunk 5]
```diff
@@ -87,37 +77,107 @@ def run(self, edit):
...
+    difftxt = u"".join(line for line in diff)
 
-        v.run_command('append', {'characters': difftxt, 'disable_tab_translation': True})
+    if difftxt == "":
+        sublime.status_message("No changes")
+        return
 
-        if not use_buffer:
-            win.run_command("show_panel", {"panel": "output.unsaved_changes"})
+    use_buffer = not view or view.settings().get('diff_changes_to_buffer')
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py]: changed to use a `diff_tabs_to_buffer` setting. Do we want to set a default for it as part of this package?

File: Diff/diff.py [hunk 6]
```diff
@@ -87,37 +77,118 @@ def run(self, edit):
...
-        difftxt = u"".join(line for line in diff)
+        show_diff_output(diff, self.view, self.view.window(), "Unsaved Changes: " + os.path.basename(self.view.file_name()), 'unsaved_changes')
 
-        if difftxt == "":
-            sublime.status_message("No changes")
-            return
+    def is_enabled(self):
+        return self.view.is_dirty() and self.view.file_name() is not None
 
-        use_buffer = self.view.settings().get('diff_changes_to_buffer')
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py:104]: fixed

File: Diff/diff.py [hunk 7]
```diff
@@ -87,37 +77,119 @@ def run(self, edit):
...
+        return self.is_visible(**kwargs)
+
+    def is_visible(self, **kwargs):
+        views = get_views_from_tab_context(self.view, **kwargs)
+        return len(views) == 2
+
+    def description(self, **kwargs):
+        selected_views = list(get_selected_views(self.view.window()))
+        if len(selected_views) == 2:
+            return 'Diff selected tabs...'
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py]: done

File: Diff/diff.py [hunk 8]
```diff
@@ -87,37 +77,119 @@ def run(self, edit):
...
+
+        return 'Diff with current tab...'
```
[the remark this answers is not in the document]
Comment (keith-hall) [reply Diff/diff.py]: done

File: Diff/diff.py [hunk 9]
```diff
@@ -87,37 +77,119 @@ def run(self, edit):
...
+    if use_buffer:
+        v = win.new_file()
+        v.set_name(name)
+        v.set_scratch(True)
+    else:
+        v = win.create_output_panel(panel_name)
+        if view:
+            v.settings().set('word_wrap', view.settings().get('word_wrap'))
+
+    v.assign_syntax('Packages/Diff/Diff.sublime-syntax')
```
[the remark this answers is not in the document]
Comment (deathaxe) [reply Diff/diff.py:113]: What's the point?

**Review by michaelblyons — APPROVED** [review]

**Pull request merged**